### PR TITLE
Implement KEMTLS based on Delegated Credentials

### PR DESCRIPTION
**Pull request by thomwiggers**
This PR implements [KEMTLS](https://ia.cr/2020/534) based on @claucece's DC branch (#28). There are a few minor fixes included in the DC stuff, surrounding the negotiation of DC supported algorithms.

This branch is targeted at the `cf` because that makes this PR reviewable: otherwise all of `circl` is included on top of my changes to the DC implementation and `crypto/tls` in general.

The following is currently implemented
* [x] KEMTLS
* [x] `crypto/kem` ergonomic KEM api
  * This package implements:
    * Kem25519: a KEM-variant of X25519. Useful for testing, but should be deleted before merge.
    * SIKEp434 provided via `circl`
    * Kyber512 provided via `circl`
* [ ] KEMTLS client authentication

## New TLS options
New supported signature algorithms (for DC):
  * [x] KEMTLSwithSIKEp434
  * [x] KEMTLSwithKyber512

New supported keyshare types (these can be negotiated independently in TLS 1.3):
* [x] SIKEp434
* [x] Kyber512

## Testing
* [ ] Unit tests
   * I've currently tested it by hard-coding the type of DC that the DC tests use. Unfortunately, those tests still need to be refactored a bit before it's easy to add more types of supported DC — but I'm sure that was on @claucece's list :smiley: 

## Stuff to do still:
* Expand the number of KEMs based on newer versions of Circl (~Kyber, hopefully, soon~)

### CODE REVIEW

**Comment by thomwiggers** [comment]
People who may be interested: @claucece, @bwesterb, @chris-wood

File: src/crypto/x509/x509.go [hunk 1]
```diff
@@ -108,6 +109,12 @@ func marshalPublicKey(pub interface{}) (publicKeyBytes []byte, publicKeyAlgorith
 		}
 		publicKeyBytes, _ = pub.MarshalBinary()
 		publicKeyAlgorithm.Algorithm = scheme.Oid()
+	case *kem.PublicKey:
+		publicKeyBytes = kem.MarshalPublicKey(pub)
+		publicKeyAlgorithm.Algorithm = oidPublicKeyKEMTLS
+	case kem.PublicKey:
+		publicKeyBytes = kem.MarshalPublicKey(&pub)
+		publicKeyAlgorithm.Algorithm = oidPublicKeyKEMTLS
```
Comment (thomwiggers) [inline src/crypto/x509/x509.go:117]: This should be simplified, but I'm having some trouble with the conventions on when to use a pointer or struct.

File: src/crypto/kem/kem.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	PublicKey []byte
+}
+
+// MarshalPublicKey produces the public key in bytes for the network
+func MarshalPublicKey(pubKey *PublicKey) []byte {
+	buf := make([]byte, 4+len(pubKey.PublicKey))
+	binary.LittleEndian.PutUint16(buf, uint16(pubKey.Id))
+	copy(buf[4:], pubKey.PublicKey)
+	return buf
+}
```
Comment (thomwiggers) [inline src/crypto/kem/kem.go:51]: Should maybe be done via https://golang.org/pkg/encoding/#BinaryMarshaler

**Comment by claucece** [comment]
Hi, @thomwiggers ! This is great! I have some time this week.. so I'll review and also update the DC branch. I was going to add DC support for client auth if it is useful for you as well. Let me know ;)

**Comment by thomwiggers** [comment]
That would be great for KEMTLS client auth as well

File: src/crypto/tls/handshake_client.go [hunk 3]
```diff
@@ -161,6 +161,9 @@ func (c *Conn) makeClientHello() (*clientHelloMsg, []clientKeysharePrivate, erro
 		}
 
 		hello.keyShares = keyShares
+
+		// only for TLS 1.3
+		hello.delegatedCredentialSupported = config.SupportDelegatedCredential
```
Comment (claucece) [inline src/crypto/tls/handshake_client.go:164]: I added this to the dc branch as well.

File: src/crypto/tls/handshake_messages.go [hunk 4]
```diff
@@ -548,7 +548,7 @@ func (m *clientHelloMsg) unmarshal(data []byte) bool {
 				if !sigAndAlgs.ReadUint16(&sigAndAlg) {
 					return false
 				}
-				m.supportedSignatureAlgorithms = append(
+				m.supportedSignatureAlgorithmsDC = append(
```
Comment (claucece) [inline src/crypto/tls/handshake_messages.go:551]: Added as well.

File: src/crypto/tls/handshake_messages.go [hunk 5]
```diff
@@ -208,7 +208,7 @@ func (m *clientHelloMsg) marshal() []byte {
 					b.AddUint16(extensionDelegatedCredentials)
 					b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
 						b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
-							for _, sigAlgo := range m.supportedSignatureAlgorithms {
+							for _, sigAlgo := range m.supportedSignatureAlgorithmsDC {
```
Comment (claucece) [inline src/crypto/tls/handshake_messages.go:211]: Added as well.

File: src/crypto/tls/common.go [hunk 6]
```diff
@@ -191,6 +191,8 @@ var supportedSignatureAlgorithms = []SignatureScheme{
 	ECDSAWithSHA1,
 }
 
+var supportedSignatureAlgorithmsDC = supportedSignatureAlgorithms
```
Comment (claucece) [inline src/crypto/tls/common.go]: This one here, I'm not really sure. Let me think about it.

**Review by claucece — CHANGES_REQUESTED** [review]

File: src/crypto/tls/common.go (hunk 6 above)
Comment (thomwiggers) [reply src/crypto/tls/common.go]: It was just something to get stuff working here — I'll definitely defer to your implementation when we rebase.

**Comment by claucece** [comment]
@thomwiggers I'm starting a new branch with this changes and refactors, as this will be too difficult to merge.

**Comment by thomwiggers** [comment]
Yeah, I'd wondered if that would be necessary.

**Comment by thomwiggers** [comment]
Closing in favour of #44

**Pull request closed without merging**